blogposts/2020/performance-improvements-in-precise-code-intel.md

```diff
@@ -0,0 +1,283 @@
+
+![eric's general commit message quality](https://storage.googleapis.com/sourcegraph-assets/efritz-commit-log.png)
+
+--- 
+
+## Fiddle with the architecture
+
+This section describes changes that were made at the architectural level, which alters which services communicate with what, what data is owned by what service, and what responsibilities services have.
+
+### Collapse network boundaries
+
+**PRs**:
+- [codeintel: Call handleEnqueue from lsifserver proxy (#10871)](https://github.com/sourcegraph/sourcegraph/pull/10871)
+- [codeintel: Call query methods from from lsifserver client (#10872)](https://github.com/sourcegraph/sourcegraph/pull/10872)
+
+Before the port to Go, there were three services written in TypeScript composing the precise code intel backend:
+
+1. _precise-code-intel-api-server_, which is the interface to the precise-code-intel database records,
+2. _precise-code-intel-bundle-manager_, which is the interface directly above SQLite bundles, and
+3. _precise-code-intel-worker_, which converts raw LSIF input into a SQLite file readable by the bundle manager.
+
+After the port to Go, the API server became a pretty thin wrapper over queries to the bundle manager. Regardless of the performance outcome, this service was destined to be removed in favor of directly querying the bundle manager from the frontend (the only consumer of the API).
+
+In order to do this with a minimal diff, we replace the network calls to the api-server by [calling the HTTP handlers directly](https://github.com/sourcegraph/sourcegraph/pull/10872/files#diff-7c26d1b2bd39d2fa821872a65e774652R40) with a fake HTTP request. This allowed us to collapse the network boundary without touching much of the api-server client, or the HTTP handler code in the api-server. This is a pretty neat trick!
+
+Additional steps were required to reduce these layers into a direct function call, but the remaining steps were only to increase the hygiene of the code and were not huge performance boosters.
+
+## Increase parallelism
+
+This section describes a handful of changes that were made to split work so that parts of it can be done concurrently (switching to another pending task when the current task is blocked) or in parallel (performed at the same time on physical different cores).
+
+### Efficient input streaming
+
+**PRs**:
+- [codeintel: Refactor lsif input parsing (#10935)](https://github.com/sourcegraph/sourcegraph/pull/10935)
+- [codeintel: LSIF line reader (#10990)](https://github.com/sourcegraph/sourcegraph/pull/10990)
+
+Even after decreasing the CPU time and allocation pressure due to JSON parsing, it remained the bottleneck. The lines fed into the correlator process were consumed almost immediately, making this an issue of a slow-producer/fast-consumer (which is much more manageable problem than a slow consumer).
```

Comment:
Did you ever take a look at https://github.com/minio/simdjson-go?

Reply:
I have not, but it doesn't have an equivalent API so the change wouldn't have been as easy. I will still look into it, but it seems like you need to manually deserialize objects via advance/iter.